indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentAllocationQueue.java

```diff
@@ -0,0 +1,685 @@
+    );
+
+    final List<SegmentAllocateRequest> successfulRequests = new ArrayList<>();
+    for (int i = 0; i < requests.size(); ++i) {
+      SegmentAllocateRequest request = requests.get(i);
+      SegmentAllocateResult result = results.get(i);
+      if (result.isSuccess()) {
+        successfulRequests.add(request);
+      }
+
+      requestBatch.handleResult(result, request);
+    }
+
+    return successfulRequests;
+  }
+
+  private Map<Interval, List<SegmentAllocateRequest>> getRequestsByInterval(
+      Set<SegmentAllocateRequest> requests,
+      Granularity tryGranularity
+  )
+  {
+    final Map<Interval, List<SegmentAllocateRequest>> tryIntervalToRequests = new HashMap<>();
+    for (SegmentAllocateRequest request : requests) {
+      Interval tryInterval = tryGranularity.bucket(request.getAction().getTimestamp());
+      if (tryInterval.contains(request.getRowInterval())) {
+        tryIntervalToRequests.computeIfAbsent(tryInterval, i -> new ArrayList<>()).add(request);
+      }
+    }
+    return tryIntervalToRequests;
+  }
+
+  private void emitTaskMetric(String metric, long value, SegmentAllocateRequest request)
+  {
+    final ServiceMetricEvent.Builder metricBuilder = ServiceMetricEvent.builder();
+    IndexTaskUtils.setTaskDimensions(metricBuilder, request.getTask());
+    metricBuilder.setDimension("taskActionType", SegmentAllocateAction.TYPE);
+    emitter.emit(metricBuilder.build(metric, value));
+  }
+
+  private void emitBatchMetric(String metric, long value, AllocateRequestKey key)
+  {
+    final ServiceMetricEvent.Builder metricBuilder = ServiceMetricEvent.builder();
+    metricBuilder.setDimension("taskActionType", SegmentAllocateAction.TYPE);
+    metricBuilder.setDimension(DruidMetrics.DATASOURCE, key.dataSource);
+    metricBuilder.setDimension(DruidMetrics.INTERVAL, key.preferredAllocationInterval.toString());
+    emitter.emit(metricBuilder.build(metric, value));
+  }
+
+  public void becomeLeader()
+  {
+    if (isEnabled()) {
+      // Start polling the queue
+      log.info("Elected leader. Starting queue processing.");
+      scheduleQueuePoll(maxWaitTimeMillis);
+    } else {
+      log.info("Elected leader but batched segment allocation is disabled.");
+    }
+  }
+
+  public void stopBeingLeader()
+  {
+    log.info("Not leader anymore. Stopping queue processing.");
+  }
+
+  /**
+   * A batch of segment allocation requests.
+   */
+  private class AllocateRequestBatch
+  {
+    private long queueTimeMillis;
+    private final AllocateRequestKey key;
+
+    /**
+     * This must be accessed through methods synchronized on this batch.
+     * It is to avoid races between a new request being added just when the batch
+     * is being processed.
+     */
+    private final Map<SegmentAllocateRequest, CompletableFuture<SegmentIdWithShardSpec>>
+        requestToFuture = new HashMap<>();
+
+    AllocateRequestBatch(AllocateRequestKey key)
+    {
+      log.info("Creating a new batch with key: %s", key);
```

Comment:
how much logging does it result in?

Reply:
Shouldn't be a lot. If there are too many tasks running, chances are they would end up sharing a batch.
This log information could be useful when not looking at metrics and trying to verify if batches are getting added to the queue. But I guess it can be a debug log line.

Maybe I should just add a log line in the processing method which prints the current queue size.

Reply:
Updated these logs to be less verbose.